src/groupchat.c

```diff
@@ -756,3 +1060,166 @@ static ToxWindow *new_group_chat(uint32_t groupnum)
 
     return ret;
 }
+
+#ifdef AUDIO
+
+#define GROUPAV_SAMPLE_RATE 48000
+#define GROUPAV_FRAME_DURATION 20
+#define GROUPAV_AUDIO_CHANNELS 1
+#define GROUPAV_SAMPLES_PER_FRAME (GROUPAV_SAMPLE_RATE * GROUPAV_FRAME_DURATION / 1000)
+
+void audio_group_callback(void *tox, uint32_t groupnumber, uint32_t peernumber, const int16_t *pcm,
+                          unsigned int samples, uint8_t channels, uint32_t sample_rate, void *userdata)
+{
+    const GroupChat *chat = &groupchats[groupnumber];
+
+    if (!chat->active) {
+        return;
+    }
+
+    for (uint32_t i = 0; i < chat->max_idx; ++i) {
+        GroupPeer *peer = &chat->peer_list[i];
+
+        if (!peer->active || peer->peernumber != peernumber) {
+            continue;
+        }
+
+        if (!peer->sending_audio) {
+            if (open_output_device(&peer->audio_out_idx,
+                                   sample_rate, GROUPAV_FRAME_DURATION, channels) != de_None) {
+                // TODO: error message?
+                return;
+            }
+
+            peer->sending_audio = true;
+
+            set_peer_audio_position(tox, groupnumber, i);
+        }
+
+        write_out(peer->audio_out_idx, pcm, samples, channels, sample_rate);
+
+        peer->last_audio_time = get_unix_time();
+
+        return;
```

Comment:
All of this code runs exactly 0 or 1 times per call to this function. It's a bit unclear because it's part of the loop body. If you factor out the "find the peer" bit into a function, then the "act on the peer we meant" bit can be behind an `if`. You could also handle the case of "no peer found", if there is something useful to do there (perhaps assert if that can't happen? Or else just return if it's normal).